### PR TITLE
[Requirements] Add vulnerability check to CI and fix vulnerabilities

### PR DESCRIPTION
**Pull request by Hedingber**
Following https://github.com/mlrun/mlrun/pull/1439 I added [safety](https://github.com/pyupio/safety) check to our CI

We had 4 vulnerabilities:
* FastAPI - fixed by bumping from 0.67.x to 0.78.x - there was a backwards incompatible change on the way in [0.74.0](https://github.com/tiangolo/fastapi/releases/tag/0.74.0) but our dependency with yield is already wrapped with `try...finally...` so we're good
* MLRun - the actual vulnerability is tensorflow 2.4.1 - newer tensorflow versions are not compatible with our CUDA and rapids versions so we can't upgrade it currently
* Numpy - we need >=1.22.2 - waiting for [Storey](https://github.com/mlrun/storey/) to remove the <1.22 limitation first and then will fix it on our side
* Kubernetes - no fix yet (on their side)

Here's the full report:
```
+==============================================================================+
|                                                                              |
|                               /$$$$$$            /$$                         |
|                              /$$__  $$          | $$                         |
|           /$$$$$$$  /$$$$$$ | $$  \__//$$$$$$  /$$$$$$   /$$   /$$           |
|          /$$_____/ |____  $$| $$$$   /$$__  $$|_  $$_/  | $$  | $$           |
|         |  $$$$$$   /$$$$$$$| $$_/  | $$$$$$$$  | $$    | $$  | $$           |
|          \____  $$ /$$__  $$| $$    | $$_____/  | $$ /$$| $$  | $$           |
|          /$$$$$$$/|  $$$$$$$| $$    |  $$$$$$$  |  $$$$/|  $$$$$$$           |
|         |_______/  \_______/|__/     \_______/   \___/   \____  $$           |
|                                                          /$$  | $$           |
|                                                         |  $$$$$$/           |
|  by pyup.io                                              \______/            |
|                                                                              |
+==============================================================================+
| REPORT                                                                       |
| checked 194 packages, using free DB (updated once a month)                   |
+============================+===========+==========================+==========+
| package                    | installed | affected                 | ID       |
+============================+===========+==========================+==========+
| numpy                      | 1.21.6    | <1.22.0                  | 44716    |
+==============================================================================+
| Numpy 1.22.0 includes a fix for CVE-2021-41496: Buffer overflow in the       |
| array_from_pyobj function of fortranobject.c, which allows attackers to      |
| conduct a Denial of Service attacks by carefully constructing an array with  |
| negative values.                                                             |
| NOTE: The vendor does not agree this is a vulnerability; the negative        |
| dimensions can only be created by an already privileged user (or             |
| internally).                                                                 |
| https://github.com/numpy/numpy/issues/19000                                  |
+==============================================================================+
| numpy                      | 1.21.6    | <1.22.0                  | 44717    |
+==============================================================================+
| Numpy 1.22.0 includes a fix for CVE-2021-34141: An incomplete string         |
| comparison in the numpy.core component in NumPy before 1.22.0 allows         |
| attackers to trigger slightly incorrect copying by constructing specific     |
| string objects. NOTE: the vendor states that this reported code behavior is  |
| "completely harmless."                                                       |
| https://github.com/numpy/numpy/issues/18993                                  |
+==============================================================================+
| numpy                      | 1.21.6    | <1.22.2                  | 44715    |
+==============================================================================+
| Numpy 1.22.2  includes a fix for CVE-2021-41495: Null Pointer Dereference    |
| vulnerability exists in numpy.sort in NumPy in the PyArray_DescrNew function |
| due to missing return-value validation, which allows attackers to conduct    |
| DoS attacks by repetitively creating sort arrays.                            |
| NOTE: While correct that validation is missing, an error can only occur due  |
| to an exhaustion of memory. If the user can exhaust memory, they are already |
| privileged. Further, it should be practically impossible to construct an     |
| attack which can target the memory exhaustion to occur at exactly this       |
| place.                                                                       |
| https://github.com/numpy/numpy/issues/19038                                  |
+==============================================================================+
| mlrun                      | 1.0.3rc2  | <=1.1.0rc1               | 48250    |
+==============================================================================+
| Mlrun 1.1.0rc1 and prior uses a version of 'TensorFlow' (2.4.1) that has     |
| known vulnerabilities.                                                       |
+==============================================================================+
| kubernetes                 | 12.0.1    | >0                       | 45114    |
+==============================================================================+
| Kubernetes (python client) uses Kubernetes API, which has an unfixed         |
| vulnerability, CVE-2021-29923: Go before 1.17 does not properly consider     |
| extraneous zero characters at the beginning of an IP address octet, which    |
| (in some situations) allows attackers to bypass access control that is based |
| on IP addresses, because of unexpected octal interpretation. This affects    |
| net.ParseIP and net.ParseCIDR. Kubernetes interprets leading zeros on IPv4   |
| addresses as decimal to keep backwards compatibility, but users relying on   |
| parser alignment will be impacted by this CVE.                               |
| https://github.com/kubernetes/kubernetes/pull/104368                         |
| https://github.com/kubernetes/kubernetes/issues/108074                       |
+==============================================================================+
| fastapi                    | 0.67.0    | <0.75.2                  | 48160    |
+==============================================================================+
| Fastapi 0.75.2 updates its NPM dependency 'swagger-ui' to include security   |
| fixes.                                                                       |
+==============================================================================+
| fastapi                    | 0.67.0    | <0.75.2                  | 48161    |
+==============================================================================+
| Fastapi 0.75.2 updates its NPM dependency 'swagger-ui' to include security   |
| fixes.                                                                       |
+==============================================================================+
| fastapi                    | 0.67.0    | <0.75.2                  | 48159    |
+==============================================================================+
| Fastapi 0.75.2 updates its dependency 'ujson' ranges to include a security   |
| fix.                                                                         |
+==============================================================================+
```

Related to https://jira.iguazeng.com/browse/ML-2252

### CODE REVIEW

File: automation/package_test/test.py [hunk 1]
```diff
@@ -89,7 +97,9 @@ def _run_test(self, test_function, extra, results, test_key):
         try:
             test_function(extra)
         except Exception:
-            results[extra][test_key]["passed"] = False
+            results[extra].setdefault(test_key, {})["passed"] = False
+        else:
+            results[extra].setdefault(test_key, {})["passed"] = True
```
Comment (Tankilevitch) [inline automation/package_test/test.py:102]: I guess you thought about it but why using the else and not just passing the `["passed"] = True` after the `test_function(extra)`

File: tests/automation/package_test/test_package_test.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            "output": "",
+        },
+    ]
+    for case in cases:
+        logger.info("Testing case", case=case)
+
+        def _run_command_mock(command, *args, **kwargs):
+            if command == "pip install safety":
+                return 0, "", ""
```
Comment (Tankilevitch) [inline tests/automation/package_test/test_package_test.py:44]:
```suggestion
            # expected to be the first command to be executed in _run_command through _test_requirements_vulnerabilities
            if command == "pip install safety":
                return 0, "", ""
```

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Great PR description, added some comments mainly on inline docstring / comments missing which will be use full for new readers.

File: automation/package_test/test.py (hunk 1 above)
Comment (Hedingber) [reply automation/package_test/test.py:102]: I actually didn't, but in general when you wrap with try except you want in the try only the code you expect to fail and want to catch error for, here it's only the `test_function` call, if for example the `extra` key is missing in the `results` dict meaning a `KeyError` will be thrown I don't want to catch that

File: automation/package_test/test.py [hunk 3]
```diff
@@ -113,6 +123,91 @@ def _test_extra_imports(self, extra):
...
+                ],
+                "numpy": [
+                    {
+                        "pattern": r"(.*)",
+                        "reason": "Storey require numpy <1.22 and the vulnerabilities fixed after this version, so "
+                        "we're waiting for Storey to release a version that supports the newer numpy and then"
+                        " we'll fix it as well",
+                    },
+                ],
+            }
```
[the remark this answers is not in the document]
Comment (Hedingber) [reply automation/package_test/test.py:183]: TBH I feel like the code is pretty much self explanatory, if there will be a new vulnerability the test will fail, so either you fix it, or you ignore it (by adding to the ignored list)
Can you give some suggestion to what kind of comment you'll add here ?

**Comment by nicklausbrown** [comment]
With the CUDA version, is that in reference to the docker images associated with mlrun?

File: automation/package_test/test.py (hunk 1 above)
Comment (Tankilevitch) [reply automation/package_test/test.py:102]: Sababa agreed

File: automation/package_test/test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply automation/package_test/test.py:183]: The code is self-explanatory; what I think could be useful are descriptions / or some guidelines for the developer / maintainer that in his CI this test failed, what he should strive for (try to fix, if unable explain why, if something has already been fixed then remove, I guess we don't want to just leave all the history of the security checks that failed previously).

**Review by Tankilevitch — APPROVED** [review]

**Pull request merged**